core/src/main/java/org/pentaho/di/compatibility/Value.java

```diff
@@ -0,0 +1,3960 @@
+      } else {
+        throw new KettleValueException( "Function NUM2DAT only works on a number" );
+      }
+    }
+    return this;
+  }
+
+  public Value str2dat( String arg0 ) throws KettleValueException {
+    return str2dat( arg0, null );
+  }
+
+  public Value str2dat( String arg0, String arg1 ) throws KettleValueException {
+    if ( isNull() ) {
+      setType( VALUE_TYPE_DATE );
+    } else {
+      // System.out.println("Convert string ["+string+"] to date using pattern '"+arg0+"'");
+
+      SimpleDateFormat df = new SimpleDateFormat();
+
+      DateFormatSymbols dfs = new DateFormatSymbols();
+      if ( arg1 != null ) {
+        dfs.setLocalPatternChars( arg1 );
+      }
+      if ( arg0 != null ) {
+        df.applyPattern( arg0 );
+      }
+
+      try {
+        value.setDate( df.parse( getString() ) );
+        setType( VALUE_TYPE_DATE );
+        setLength( -1, -1 );
+      } catch ( Exception e ) {
+        setType( VALUE_TYPE_DATE );
+        setNull();
+        throw new KettleValueException( "TO_DATE Couldn't convert String to Date" + e.toString() );
+      }
+    }
+    return this;
+  }
+
+  // implement the TO_NUMBER function, arguments in args[]
+  public Value str2num() throws KettleValueException {
+    return str2num( null, null, null, null );
+  }
+
+  public Value str2num( String pattern ) throws KettleValueException {
+    return str2num( pattern, null, null, null );
+  }
+
+  public Value str2num( String pattern, String decimal ) throws KettleValueException {
+    return str2num( pattern, decimal, null, null );
+  }
+
+  public Value str2num( String pattern, String decimal, String grouping ) throws KettleValueException {
+    return str2num( pattern, decimal, grouping, null );
+  }
+
+  public Value str2num( String pattern, String decimal, String grouping, String currency ) throws KettleValueException {
+    // 0 : pattern
+    // 1 : Decimal separator
+    // 2 : Grouping separator
+    // 3 : Currency symbol
+
+    if ( isNull() ) {
+      setType( VALUE_TYPE_STRING );
+    } else {
+      if ( getType() == VALUE_TYPE_STRING ) {
+        if ( getString() == null ) {
+          setNull();
+          setValue( 0.0 );
+        } else {
+          NumberFormat nf = NumberFormat.getInstance();
+          DecimalFormat df = (DecimalFormat) nf;
+          DecimalFormatSymbols dfs = new DecimalFormatSymbols();
+
+          if ( !Utils.isEmpty( pattern ) ) {
+            df.applyPattern( pattern );
+          }
+          if ( !Utils.isEmpty( decimal ) ) {
+            dfs.setDecimalSeparator( decimal.charAt( 0 ) );
+          }
+          if ( !Utils.isEmpty( grouping ) ) {
+            dfs.setGroupingSeparator( grouping.charAt( 0 ) );
+          }
+          if ( !Utils.isEmpty( currency ) ) {
+            dfs.setCurrencySymbol( currency );
+          }
+          try {
+            df.setDecimalFormatSymbols( dfs );
+            setValue( df.parse( getString() ).doubleValue() );
+          } catch ( Exception e ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)